### PR TITLE
[orchagent][neighorch] traffic convergence acceleration for staticroutebfd

### PR DESCRIPTION
**Pull request by baorliu**
**What I did**
valid/invalid nexthop based on BFD session state.
Microsoft ADO: https://msazure.visualstudio.com/One/_workitems/edit/17793093

**Why I did it**
to accelerate traffic convergence when BFD session state changes

**How I verified it**
load test image to testbed, set up static route nexthop group with 2 members.
trigger BFD session UP/DOWN and check sairedis.rec to see add/remove member transection:

```
2023-05-09.21:54:24.553671|C|SAI_OBJECT_TYPE_ROUTE_ENTRY||{"dest":"192.85.4.0/24","switch_id":"oid:0x21000000000000","vr":"oid:0x3000000000042"}|SAI_ROUTE_ENTRY_ATTR_NEXT_HOP_ID=oid:0x60000000009a9||{"dest":"192.85.0.0/24","switch_id":"oid:0x21000000000000","vr":"oid:0x3000000000042"}|SAI_ROUTE_ENTRY_ATTR_NEXT_HOP_ID=oid:0x60000000009a8

...

2023-05-09.21:54:24.558581|c|SAI_OBJECT_TYPE_NEIGHBOR_ENTRY:{"ip":"192.85.0.24","rif":"oid:0x60000000009a8","switch_id":"oid:0x21000000000000"}|SAI_NEIGHBOR_ENTRY_ATTR_DST_MAC_ADDRESS=E8:EB:34:E7:D3:E0
2023-05-09.21:54:24.559243|c|SAI_OBJECT_TYPE_NEXT_HOP:oid:0x40000000009aa|SAI_NEXT_HOP_ATTR_TYPE=SAI_NEXT_HOP_TYPE_IP|SAI_NEXT_HOP_ATTR_IP=192.85.0.24|SAI_NEXT_HOP_ATTR_ROUTER_INTERFACE_ID=oid:0x60000000009a8
2023-05-09.21:54:24.560322|c|SAI_OBJECT_TYPE_NEIGHBOR_ENTRY:{"ip":"192.85.4.24","rif":"oid:0x60000000009a9","switch_id":"oid:0x21000000000000"}|SAI_NEIGHBOR_ENTRY_ATTR_DST_MAC_ADDRESS=E8:EB:34:E7:D3:E0
2023-05-09.21:54:24.561087|c|SAI_OBJECT_TYPE_NEXT_HOP:oid:0x40000000009ab|SAI_NEXT_HOP_ATTR_TYPE=SAI_NEXT_HOP_TYPE_IP|SAI_NEXT_HOP_ATTR_IP=192.85.4.24|SAI_NEXT_HOP_ATTR_ROUTER_INTERFACE_ID=oid:0x60000000009a9
2023-05-09.21:54:24.562126|c|SAI_OBJECT_TYPE_NEXT_HOP_GROUP:oid:0x50000000009ac|SAI_NEXT_HOP_GROUP_ATTR_TYPE=SAI_NEXT_HOP_GROUP_TYPE_DYNAMIC_UNORDERED_ECMP
2023-05-09.21:54:24.562766|C|SAI_OBJECT_TYPE_NEXT_HOP_GROUP_MEMBER||oid:0x2d0000000009ad|SAI_NEXT_HOP_GROUP_MEMBER_ATTR_NEXT_HOP_GROUP_ID=oid:0x50000000009ac|SAI_NEXT_HOP_GROUP_MEMBER_ATTR_NEXT_HOP_ID=oid:0x40000000009aa|SAI_NEXT_HOP_GROUP_MEMBER_ATTR_WEIGHT=1||oid:0x2d0000000009ae|SAI_NEXT_HOP_GROUP_MEMBER_ATTR_NEXT_HOP_GROUP_ID=oid:0x50000000009ac|SAI_NEXT_HOP_GROUP_MEMBER_ATTR_NEXT_HOP_ID=oid:0x40000000009ab|SAI_NEXT_HOP_GROUP_MEMBER_ATTR_WEIGHT=1
2023-05-09.21:54:24.563631|C|SAI_OBJECT_TYPE_ROUTE_ENTRY||{"dest":"2.2.2.0/24","switch_id":"oid:0x21000000000000","vr":"oid:0x3000000000042"}|SAI_ROUTE_ENTRY_ATTR_NEXT_HOP_ID=oid:0x50000000009ac

...
2023-05-09.21:58:53.518641|c|SAI_OBJECT_TYPE_BFD_SESSION:oid:0x450000000009af|SAI_BFD_SESSION_ATTR_TYPE=SAI_BFD_SESSION_TYPE_ASYNC_ACTIVE|SAI_BFD_SESSION_ATTR_LOCAL_DISCRIMINATOR=1|SAI_BFD_SESSION_ATTR_UDP_SRC_PORT=49152|SAI_BFD_SESSION_ATTR_REMOTE_DISCRIMINATOR=0|SAI_BFD_SESSION_ATTR_BFD_ENCAPSULATION_TYPE=SAI_BFD_ENCAPSULATION_TYPE_NONE|SAI_BFD_SESSION_ATTR_IPHDR_VERSION=4|SAI_BFD_SESSION_ATTR_SRC_IP_ADDRESS=192.85.0.29|SAI_BFD_SESSION_ATTR_DST_IP_ADDRESS=192.85.0.24|SAI_BFD_SESSION_ATTR_MIN_TX=1000000|SAI_BFD_SESSION_ATTR_MIN_RX=1000000|SAI_BFD_SESSION_ATTR_MULTIPLIER=10|SAI_BFD_SESSION_ATTR_TOS=192|SAI_BFD_SESSION_ATTR_MULTIHOP=true|SAI_BFD_SESSION_ATTR_VIRTUAL_ROUTER=oid:0x3000000000042
...
2023-05-09.21:59:59.529446|n|bfd_session_state_change|[{"bfd_session_id":"oid:0x450000000009af","session_state":"SAI_BFD_SESSION_STATE_UP"}]|
2023-05-09.21:59:59.529841|c|SAI_OBJECT_TYPE_NEXT_HOP_GROUP_MEMBER:oid:0x2d0000000009b0|SAI_NEXT_HOP_GROUP_MEMBER_ATTR_NEXT_HOP_GROUP_ID=oid:0x50000000009ac|SAI_NEXT_HOP_GROUP_MEMBER_ATTR_NEXT_HOP_ID=oid:0x40000000009aa|SAI_NEXT_HOP_GROUP_MEMBER_ATTR_WEIGHT=1
2023-05-09.22:02:25.563897|n|bfd_session_state_change|[{"bfd_session_id":"oid:0x450000000009af","session_state":"SAI_BFD_SESSION_STATE_DOWN"}]|
2023-05-09.22:02:25.564160|r|SAI_OBJECT_TYPE_NEXT_HOP_GROUP_MEMBER:oid:0x2d0000000009b0

```

**Details if related**

### CODE REVIEW

File: orchagent/neighorch.cpp [hunk 1]
```diff
@@ -433,6 +441,66 @@ bool NeighOrch::ifChangeInformNextHop(const string &alias, bool if_up)
...
+    {
+        if (nhop->first.ip_address != peer_address)
+        {
+            continue;
+        }
+
+        if (state == SAI_BFD_SESSION_STATE_UP)
+        {
+            SWSS_LOG_INFO("updateNextHop get BFD session UP event, key %s", key.c_str());
+            rc = clearNextHopFlag(nhop->first, NHFLAGS_IFDOWN);
```
Comment (prsunny) [inline orchagent/neighorch.cpp:485]: So this is going to update all routes that has this nexthop, irrespective of whether BFD is set to true for such route or not, correct?

**Review by prsunny — COMMENTED** [review]
Please add test coverage

File: orchagent/neighorch.cpp (hunk 1 above)
Comment (baorliu) [reply orchagent/neighorch.cpp:485]: Yes. From BFD SESSION itself (or BFD state event), it cannot tell if it is created by StaticRouteBfd or not.

Comment (prsunny) [reply orchagent/neighorch.cpp:485]: @abdosi for viz

File: orchagent/neighorch.cpp [hunk 2]
```diff
@@ -433,6 +441,66 @@ bool NeighOrch::ifChangeInformNextHop(const string &alias, bool if_up)
...
+        }
+        else
+        {
+            SWSS_LOG_INFO("updateNextHop get BFD session DOWN event, key %s", key.c_str());
+            rc = setNextHopFlag(nhop->first, NHFLAGS_IFDOWN);
+        }
+
+        if (rc == true)
```
[the remark this answers is not in the document]
Comment (baorliu) [reply orchagent/neighorch.cpp:493]: copied from interface up/down handling. I will update it in the next commit.

**Review by abdosi — COMMENTED** [review]
Please add UT. rest looks good

**Review by prsunny — APPROVED** [review]

**Review by prsunny — APPROVED** [review]

**Review by abdosi — APPROVED** [review]

**Comment by baorliu** [comment]
> 

I checked the log error message, I did not see any connection between this PR and the above UT test failure.
but when I debugged the UT for this PR, I did see that some timing related failure. For example, using redis API to update redis table, and using another API to check result. The checking fails sometime if there is no delay (or delay is not long enough) between the 2 API calls.
The first error happened before new added test code, test_nhg.py:947. new added test code starts at test_nhg.py:990, inside same function.

```
2023-05-26T23:49:40.2891692Z =================================== FAILURES ===================================
2023-05-26T23:49:40.2892540Z ____________________ TestNextHopGroup.test_route_nhg[false] ____________________
2023-05-26T23:49:40.2893014Z 
2023-05-26T23:49:40.2893674Z self = <test_nhg.TestNextHopGroup object at 0x7f2b720f1340>
2023-05-26T23:49:40.2896182Z ordered_ecmp = 'false'
2023-05-26T23:49:40.2896517Z dvs = <conftest.DockerVirtualSwitch object at 0x7f2b723017c0>
2023-05-26T23:49:40.2896877Z dvs_route = <dvslib.dvs_route.DVSRoute object at 0x7f2b72103fd0>
2023-05-26T23:49:40.2897214Z testlog = <function testlog at 0x7f2b722c9b80>
2023-05-26T23:49:40.2897372Z 
2023-05-26T23:49:40.2897837Z     @pytest.mark.parametrize('ordered_ecmp', ['false', 'true'])
2023-05-26T23:49:40.2898220Z     def test_route_nhg(self, ordered_ecmp, dvs, dvs_route, testlog):
2023-05-26T23:49:40.2898551Z         self.init_test(dvs, 3)
2023-05-26T23:49:40.2898867Z         nhip_seqid_map = {"10.0.0.1" : "1", "10.0.0.3" : "2" , "10.0.0.5" : "3" }
2023-05-26T23:49:40.2899140Z     
2023-05-26T23:49:40.2899489Z         if ordered_ecmp == 'true':
2023-05-26T23:49:40.2899771Z            self.enable_ordered_ecmp()
2023-05-26T23:49:40.2900029Z     
2023-05-26T23:49:40.2900259Z         rtprefix = "2.2.2.0/24"
2023-05-26T23:49:40.2900479Z     
2023-05-26T23:49:40.2900761Z >       dvs_route.check_asicdb_deleted_route_entries([rtprefix])
2023-05-26T23:49:40.2900919Z 
2023-05-26T23:49:40.2901156Z test_nhg.py:947:

**Comment by baorliu** [comment]
> 

I do see an error from the above link. I am trying to find why it can pass the master branch pipeline.
```
2023-05-26T23:49:40.3662359Z 	<class 'AttributeError'>
2023-05-26T23:49:40.3662770Z 	'DockerVirtualSwitch' object has no attribute 'adb'

**Comment by baorliu** [comment]
> > 
> 
> I do see an error from the above link. I am trying to find why it can pass the master branch pipeline.
> 
> ```
> 2023-05-26T23:49:40.3661861Z test_route_nhg[false] failed (1 runs remaining out of 2).
> 2023-05-26T23:49:40.3662359Z 	<class 'AttributeError'>
> 2023-05-26T23:49:40.3662770Z 	'DockerVirtualSwitch' object has no attribute 'adb'
> 2023-05-26T23:49:40.3663149Z 	[<TracebackEntry /agent/_work/1/s/tests/test_nhg.py:1015>, <TracebackEntry /agent/_work/1/s/tests/test_nhg.py:181>]
> 
> That code dvs.adb has been in test_bfd.py for 2 years. did not expect it fails in etst_nhg.py.
> https://github.com/sonic-net/sonic-swss/blob/98d2b0c6233a90f3786d296c4b7da2bb83412359/tests/test_bfd.py#L41
>         ntf = swsscommon.NotificationProducer(dvs.adb, "NOTIFICATIONS")
> ```

I have a PR to fix this error, all the tests in test_nhg.py passed. but see some unrelated failures. 
see comments in the following PR.
https://github.com/sonic-net/sonic-swss/pull/2803

**Comment by ayurkiv-nvda** [comment]
> > > 
> > 
> > 
> > I do see an error from the above link. I am trying to find why it can pass the master branch pipeline.
> > ```
> > 2023-05-26T23:49:40.3661861Z test_route_nhg[false] failed (1 runs remaining out of 2).
> > 2023-05-26T23:49:40.3662359Z 	<class 'AttributeError'>
> > 2023-05-26T23:49:40.3662770Z 	'DockerVirtualSwitch' object has no attribute 'adb'
> > 2023-05-26T23:49:40.3663149Z 	[<TracebackEntry /agent/_work/1/s/tests/test_nhg.py:1015>, <TracebackEntry /agent/_work/1/s/tests/test_nhg.py:181>]
> > 
> > That code dvs.adb has been in test_bfd.py for 2 years. did not expect it fails in etst_nhg.py.
> > https://github.com/sonic-net/sonic-swss/blob/98d2b0c6233a90f3786d296c4b7da2bb83412359/tests/test_bfd.py#L41
> >         ntf = swsscommon.NotificationProducer(dvs.adb, "NOTIFICATIONS")
> > ```
> 
> I have a PR to fix this error, all the tests in test_nhg.py passed. but see some unrelated failures. see comments in the following PR. #2803
> 
> update: all checks are passed with rerun

This PR https://github.com/sonic-net/sonic-swss/pull/2803 will fix problem with test run failures on 202205?

**Comment by baorliu** [comment]
> 
The PR #2803 can fix test_nhg.py related issues (related changes in this PR). 
But I see that some test itself may has issue. I had 3 consecutive pipeline triggers, in the first one some tests failed (see the PR#2803 comment), the last two passed. So some tests are not stable. Even before this PR, I created another debugging PR without any logic change, I saw failure too (see the first comment in the draft PR#2774).

**Pull request merged**